### ISSUE TITLE
IPA node requests to load SDXL everytime as I load a new image as input

### ISSUE THREAD

**Issue by AlexFeYang** [cubiq/ComfyUI_IPAdapter_plus#579]
I use the --only-gpu mode on 4090,
我在 4090 上使用 --only-gpu 模式，
![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/157102924/541749c2-7df5-4c0e-beee-b1bec9fecd8c)

It requests to load SDXL everytime I load a new image as input? why?
每次我加载新图像作为输入时，它都会请求加载 SDXL？为什么？

**Comment by AlexFeYang** [comment]
alright, I am brainless here，I test with the basic workflow, it's not beacause of loading SDXL
![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/157102924/235730c0-7c3a-4bd7-9b47-0d1f50be1e4b)
but I still don't know why it takes IPA 4s to load a new image.
I used to use the diffusers directly to load image with IPA, it's won't take this much of time to load a new image as input

**Comment by cubiq** [comment]
does it do the same without --only-gpu ?

**Comment by AlexFeYang** [comment]
![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/157102924/0be84f69-d574-4b78-a838-c807a1817d3b)
I switch --highvram this time, nothing changes
still spend another 4s on ipadapter

**Comment by AlexFeYang** [comment]
![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/157102924/39154b97-d71b-4166-816f-03cb9097fb21)
still, i am using normalvarm this time.
when I change a image at image loader, it will spend another 4s

**Comment by cubiq** [comment]
can you post the image and the workflow?

**Comment by AlexFeYang** [comment]
![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/157102924/7a94e728-ad69-428b-86d8-db2bb73bdb3b)
I am using animagine

**Comment by cubiq** [comment]
can you try to pass the reference to prepimageforclipvision?

**Comment by AlexFeYang** [comment]
![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/157102924/3bd6aa1d-30a9-4997-87c5-b719d74a70be)
no help

**Comment by AlexFeYang** [comment]
![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/157102924/41f6b4ab-24f3-47b9-8b9f-0803c748c341)

**Comment by AlexFeYang** [comment]
is there anything to load inside the IPAdapter nodes? I am using cloud pod now.
my own computer, 4060 16g will take 2s to load a new image

**Comment by cubiq** [comment]
I don't know why it takes so long. on my 4090 loading one image is almost instantaneous

**Comment by AlexFeYang** [comment]
Yes, I used to work with diffusers, and it loads image within sencond,
that's why I am so comfused now.

**Comment by AlexFeYang** [comment]
This case is reproducible. Every Pod I start is like this, even in Docker. Only things here is that they all need to link to the model storage, so is there anything i should prepare of? I thought the ipa model and the clip vision and the image has been loaded ahead, so what's missing?

**Comment by AlexFeYang** [comment]
![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/157102924/9eac6b32-4973-421e-8543-fd47f4f002e3)
![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/157102924/90ffd7c6-fb97-4d0c-aa2d-d1fb3423286f)
I turn the weight from 0.8 to 0.7, it's start to load sdxl and a new model

**Comment by AlexFeYang** [comment]
![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/157102924/bfecd7a3-4b7a-4e3d-a7c2-8b858c36059c)
I add logging to your code
![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/157102924/073bd17a-b017-4c2d-960f-76740f3131f8)
it seems these codes take most of the time, any thoughts?
'''
self.image_proj_model.load_state_dict(ipadapter_model["image_proj"])
self.ip_layers = To_KV(ipadapter_model["ip_adapter"])
'''

**Comment by cubiq** [comment]
are you in fp32?

**Comment by AlexFeYang** [comment]
nope, I am using bfp16 by default, and I alse tested it with fp16, nothing helps,
![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/157102924/a6d2111a-3a36-415a-9aae-2987e22d52b6)
```
for key, value in state_dict.items():
            self.to_kvs[key.replace(".weight", "").replace(".", "_")] = nn.Linear(value.shape[1], value.shape[0], bias=False)
            self.to_kvs[key.replace(".weight", "").replace(".", "_")].weight.data = value
```
this is the code costing so much time, I don't know what to do with it.

**Comment by AlexFeYang** [comment]
Still need help sir! much appreciated!! @cubiq

**Comment by cubiq** [comment]
I still don't understand if this happens locally or on a cloud service

**Comment by AlexFeYang** [comment]
I changed the logging, it seems running on cpu, @cubiq 
![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/157102924/3a4f7a27-4783-4e50-8abc-80b97d3c1549)

![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/157102924/201ccfc4-10f0-4ec7-bc18-09f1ff419c5d)

**Comment by AlexFeYang** [comment]
I solved the problem by moving the toKV code to loader nodes, it works fine for now.@cubiq  hope everything goes well! 
Thank you for your patient help.

**Comment by cubiq** [comment]
what did you do exactly?

**Comment by AlexFeYang** [comment]
![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/157102924/8feaf48e-6402-4840-9278-30a8b4d058d2)
I move the to KV code from here to IPAdapterUnifiedLoader and IPAdapterModelLoader
![image](https://github.com/cubiq/ComfyUI_IPAdapter_plus/assets/157102924/067fa83c-4809-44fc-aa30-417b8d730ccd)
these code is running on cpu, that's why my 4090 not working well, maybe your pc have a better cpu
I test the original code on server from us to Europe, and the Europe 4090 run faster, that's because the cpu is better perhaps.
I tested it worked well so far. There seemed to be no problem. You see if there is any problem? @cubiq 
maybe there are some other init job should be place outside the core nodes to upgrade the performance.

**Comment by cubiq** [comment]
it's just a matter of moving it to the gpu before performing the operation, what I'm wondering is if it should be done automatically or not. if one has little vram it might be better to keep it on the cpu

**Comment by AlexFeYang** [comment]
you are right, I still wish to solve the problem here, I still don't get why my local machine and cloud pod all running the TO_KV function on cpu, it's seems outside this function everywhere's device is correct on GPU.

**Comment by AlexFeYang** [comment]
it will be better if you tell me where to set the related models on GPU directly, just affect the to_KV function will be ok. I am a student lack of experience on Pytorch stuff

**Comment by cubiq** [comment]
I'll make some more tests on this, thanks for bringing it out

**Comment by TmLev** [comment]
Related https://github.com/cubiq/ComfyUI_IPAdapter_plus/issues/179

**Comment by jelling** [comment]
I don't think we will see further improvements on this issue since the repo is now in maintenance mode. 

But for data, it takes approximately the same amount of time to apply saved embeds as it does to load embeds from a new image:

![Image](https://github.com/user-attachments/assets/c2584be6-b0ad-4e72-bb8f-5a140497cbb5)

And with a new image:

![Image](https://github.com/user-attachments/assets/ecf72364-2bb8-42ef-8f53-97ff704dd285)

Meaning that performance time is bottle-necked by the application, not the loading.
